### PR TITLE
We should refactor the agent cli output

### LINKED ISSUE #253: We should refactor the agent cli output
We currently have a bunch of flags in the agent command, some of them should be move to a sub command section.

```
root@byohnode02:/home/ecl# ./byoh-hostagent-linux-amd64 -h
Usage of ./byoh-hostagent-linux-amd64:
  -add_dir_header
    	If true, adds the file directory to the header
  -alsologtostderr
    	log to standard error as well as files
  -bundle-repo string
    	BYOH Bundle Repository (default "projects.registry.vmware.com")
  -cache-path string
    	Path to the local bundle cache (default ".")
  **-detect
    	Detects the current operating system**
  -downloadpath string
    	File System path to keep the downloads (default "/var/lib/byoh/bundles")
  **-install
    	Install a BYOH Bundle**
  -k8s string
    	Kubernetes version (default "1.22.1")
  -kubeconfig string
    	Paths to a kubeconfig. Only required if out-of-cluster.
  -label value
    	labels to attach to the ByoHost CR in the form labelname=labelVal for e.g. '--label site=apac --label cores=2'
  **-list-supported
    	List all supported OS, Kubernetes versions and BYOH Bundle names**
  -log_backtrace_at value
    	when logging hits line file:N, emit a stack trace
  -log_dir string
    	If non-empty, write log files in this directory
  -log_file string
    	If non-empty, use this log file
  -log_file_max_size uint
    	Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
  -logtostderr
    	log to standard error instead of files (default true)
  -metricsbindaddress string
    	metricsbindaddress is the TCP address that the controller should bind to for serving prometheus metrics.It can be set to "0" to disable the metrics serving (default ":8080")
  -namespace string
    	Namespace in the management cluster where you would like to register this host (default "default")
  -os string
    	OS. If used with install/uninstall, override os detection
  -preview-os-changes
    	Preview the install and uninstall changes for the specified OS
  -skip-installation
    	If you want to skip installation of the kubernetes component binaries
  -skip_headers
    	If true, avoid header prefixes in the log messages
  -skip_log_headers
    	If true, avoid headers when opening log files
  -stderrthreshold value
    	logs at or above this threshold go to stderr (default 2)
  -tag string
    	BYOH Bundle tag
  **-uninstall
    	Unnstall a BYOH Bundle**
  -v value
    	number for the log level verbosity
  -vmodule value
    	comma-separated list of pattern=N settings for file-filtered logging
```

When we try to list the supported bundles the output is as follow: seems it try to start the agent instead of just list the bundles.

```
root@byohnode02:/home/ecl# ./byoh-hostagent-linux-amd64 -kubeconfig management.conf -list-supported
I1118 06:19:39.662245    6236 host_registrar.go:37] Registering ByoHost
I1118 06:19:39.666148    6236 host_registrar.go:71] Add Network Info
I1118 06:19:41.827142    6236 deleg.go:130] controller-runtime/metrics "msg"="metrics server is starting to listen"  "addr"=":8080"
I1118 06:19:41.828111    6236 deleg.go:130]  "msg"="starting metrics server"  "path"="/metrics"
I1118 06:19:41.828476    6236 controller.go:178] controller/byohost "msg"="Starting EventSource" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="ByoHost" "source"={"Type":{"metadata":{"creationTimestamp":null},"spec":{},"status":{}}}
I1118 06:19:41.828570    6236 controller.go:186] controller/byohost "msg"="Starting Controller" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="ByoHost" 
I1118 06:19:41.930364    6236 controller.go:220] controller/byohost "msg"="Starting workers" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="ByoHost" "worker count"=1
I1118 06:19:41.930733    6236 host_reconciler.go:49] controller/byohost "msg"="Reconcile request received" "name"="byohnode02" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="ByoHost" 
I1118 06:19:41.934513    6236 host_reconciler.go:88] controller/byohost "msg"="Machine ref not yet set" "name"="byohnode02" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="ByoHost" 
^CI1118 06:54:30.215121    6236 controller.go:240] controller/byohost "msg"="Shutdown signal received, waiting for all workers to finish" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="ByoHost" 
I1118 06:54:30.215303    6236 controller.go:242] controller/byohost "msg"="All workers finished" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="ByoHost"
```

### PR DESCRIPTION
**Pull request by huchen2021**
Fixes #253

There are a lot of wrong flag for byoh-hostagent

1) The following flags belong to "cli", not " byoh-hostagent". It should put all flag function in main function of "cli", not as global variable.

- -bundle-repo string
- -cache-path string
- -detect
- -install
- -k8s string
- -list-supported
- -os string
- -preview-os-changes
- -tag string
- -uninstall

2) The following flags belong to klog which we don't need it anymore

- -add_dir_header
- -alsologtostderr
- -log_backtrace_at value
- -log_dir string
- -log_file string
- -log_file_max_size uint
- -logtostderr
- -skip_headers
- -skip_log_headers
- -stderrthreshold value
- -v value
- -vmodule value

After fix this,  byoh-hostagent only has the following flags:
- -downloadpath string
- -kubeconfig string
- -label value
- -metricsbindaddress string
- -namespace string
- -skip-installation

Plus, In order to avoid such a bug again, I add a test for --help command. Any unexpected option will caused the test failure.

### CODE REVIEW

**Review by anusha94 — COMMENTED** [review]
Can we write a test for `-h` or `--help` command?

**Comment by huchen2021** [comment]
> Can we write a test for `-h` or `--help` command?
Yes, it will help us to avoid such a bug again. Can we open a new issue to track it?

File: agent/installer/cli.go [hunk 1]
```diff
@@ -38,6 +38,17 @@ var (
 func Main() {
 	klogger = klogr.New()
 
+	listSupportedFlag = flag.Bool("list-supported", false, "List all supported OS, Kubernetes versions and BYOH Bundle names")
+	detectOSFlag = flag.Bool("detect", false, "Detects the current operating system")
+	installFlag = flag.Bool("install", false, "Install a BYOH Bundle")
+	uninstallFlag = flag.Bool("uninstall", false, "Unnstall a BYOH Bundle")
+	bundleRepoFlag = flag.String("bundle-repo", "projects.registry.vmware.com", "BYOH Bundle Repository")
+	cachePathFlag = flag.String("cache-path", ".", "Path to the local bundle cache")
+	k8sFlag = flag.String("k8s", "1.22.1", "Kubernetes version")
```
Comment (anusha94) [inline agent/installer/cli.go:47]: @jeuxdemains @georgievaVMW 
Why is the default version `1.22.1` here? Why are we assuming any default version at all?

Comment (jeuxdemains) [reply agent/installer/cli.go:47]: I'm not sure why we have a default option. My assumption is that this is the minimal supported one and it defaults to it if it's missing when the CLI is invoked. 
Should we remove it?

Comment (anusha94) [reply agent/installer/cli.go:47]: atm, I don't understand why we need a default version. Anyway, this is the CLI. We can discuss it on a different thread or tackle if this becomes an issue. Resolving this conversation. We are good with this PR.

**Review by anusha94 — APPROVED** [review]
lgtm

File: agent/help_flag_test.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/gexec"
+)
+
+var _ = Describe("Help flag for host agent", func() {
+	Context("When the help flag is provided", func() {
+		var (
+			expectedOptions = []string{
```
[the remark this answers is not in the document]
Comment (jeuxdemains) [reply agent/help_flag_test.go:18]: It does, there is the output of --help of the CLI:

`jeux@jeux-VirtualBox:~/cluster-api-provider-bringyourownhost-1/agent/installer/cli$ ./cli --help
Usage of ./cli:
  -bundle-repo string
    	BYOH Bundle Repository (default "projects.registry.vmware.com")
  -cache-path string
    	Path to the local bundle cache (default ".")
  -detect
    	Detects the current operating system
  -install
    	Install a BYOH Bundle
  -k8s string
    	Kubernetes version (default "1.22.1")
  -list-supported
    	List all supported OS, Kubernetes versions and BYOH Bundle names
  -os string
    	OS. If used with install/uninstall, override os detection
  -preview-os-changes
    	Preview the install and uninstall changes for the specified OS
  -tag string
    	BYOH Bundle tag
  -uninstall
    	Unnstall a BYOH Bundle

`

[the remark this answers is not in the document]
Comment (dharmjit) [reply agent/help_flag_test.go:18]: @huchen2021 As the description is again part of public API, I think it would be useful to test it as well.

**Review by dharmjit — COMMENTED** [review]
LGTM. 

Could we keep all flag-related tests in a single test file.?

**Comment by huchen2021** [comment]
> LGTM.
> 
> Could we keep all flag-related tests in a single test file.?

We already have label_flag_test.go. And If we put label_flag_test.go and help_flag_test.go together, it would be huge file. Don't think it is a good idea.

File: agent/help_flag_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (huchen2021) [reply agent/help_flag_test.go:18]: @dharmjit It's very werid to test the description. Don't think it is good idea.

[the remark this answers is not in the document]
Comment (dharmjit) [reply agent/help_flag_test.go:18]: No strong opinions, but I was thinking we should have tests to validate the description changes as well. I will leave it upto you to decide.

**Comment by dharmjit** [comment]
> We already have label_flag_test.go. And If we put label_flag_test.go and help_flag_test.go together

Again no strong opinions, but IMO we are keeping a single test file for a package, here for agent `main` we split that into 2 files `agent_test.go` and `label_flag_test.go` as the functionalities are different and it does not require envtest/k8s client, etc.

File: agent/help_flag_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (huchen2021) [reply agent/help_flag_test.go:18]: @dharmjit Actually, all options and descriptions are generated by flag library function as followed:
image

It's possible to have wrong option lists if we use it not properly. But It's almost impossible to have wrong description for one flag. Let's have confidence about flag library function.

**Review by dharmjit — APPROVED** [review]

**Pull request merged**